File: mobile-flutter/lib/models/booking.dart
/// Mirrors the web app's `BookingStatus` union (`web/src/types/booking.ts`)
/// as a Dart `enum` (idiomatic per this project's Dart pattern rules -
/// sealed/enum types with exhaustive `switch`, rather than a raw string
/// field).
enum BookingStatus {
  confirmed,
  waitlisted,
  cancelled,
  checkedIn;

  static BookingStatus fromWire(String? value) {
    return switch (value) {
      'confirmed' => BookingStatus.confirmed,
      'waitlisted' => BookingStatus.waitlisted,
      'cancelled' => BookingStatus.cancelled,
      'checked_in' => BookingStatus.checkedIn,
      // A booking status is a closed, platform-validated enum on the server
      // (see `plan/build-plan.md`) - an unrecognized wire value would only
      // ever come from a client/server version skew, never legitimate data,
      // so this falls back to the safest read: treat it as cancelled (never
      // counts toward capacity, never displayed as a valid active ticket).
      _ => BookingStatus.cancelled,
    };
  }

  String get wireValue {
    return switch (this) {
      BookingStatus.confirmed => 'confirmed',
      BookingStatus.waitlisted => 'waitlisted',
      BookingStatus.cancelled => 'cancelled',
      BookingStatus.checkedIn => 'checked_in',
    };
  }

  String get label {
    return switch (this) {
      BookingStatus.confirmed => 'Confirmed',
      BookingStatus.waitlisted => 'Waitlisted',
      BookingStatus.cancelled => 'Cancelled',
      BookingStatus.checkedIn => 'Checked in',
    };
  }
}

/// Mirrors the web app's `BookingDoc` (`web/src/types/booking.ts`) and the
/// live `bookings` collection schema (`eventId`, `userId`, `userName`,
/// `status`, `qrToken`).
class Booking {
  const Booking({
    required this.id,
    required this.eventId,
    required this.userId,
    required this.userName,
    required this.status,
    required this.qrToken,
    required this.createdAt,
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['_id'] as String? ?? '',
      eventId: json['eventId'] as String? ?? '',
      userId: json['userId'] as String? ?? '',
      userName: json['userName'] as String? ?? '',
      status: BookingStatus.fromWire(json['status'] as String?),
      qrToken: json['qrToken'] as String? ?? '',
      createdAt:
          DateTime.tryParse(json['createdAt'] as String? ?? '') ??
          DateTime.now(),
    );
  }

  final String id;
  final String eventId;
  final String userId;
  final String userName;
  final BookingStatus status;
  final String qrToken;
  final DateTime createdAt;
}

File: swift/Sources/MudbaseShowcaseEvents/Models/Booking.swift
import Foundation

/// Mirrors `web/src/types/booking.ts`'s `BookingStatus` union exactly.
enum BookingStatus: String, Sendable, Equatable, CaseIterable {
    case confirmed
    case waitlisted
    case cancelled
    case checkedIn = "checked_in"

    var label: String {
        switch self {
        case .confirmed: return "Confirmed"
        case .waitlisted: return "Waitlisted"
        case .cancelled: return "Cancelled"
        case .checkedIn: return "Checked in"
        }
    }
}

/// Mirrors `web/src/types/booking.ts`'s `BookingDoc` interface exactly.
struct Booking: Sendable, Equatable, Identifiable {
    let id: String
    let createdAt: Date?
    let eventId: String
    let userId: String
    let userName: String
    let status: BookingStatus
    let qrToken: String

    init(document: MudbaseDocument) {
        id = document.id
        createdAt = document.createdAt
        eventId = document.string("eventId") ?? ""
        userId = document.string("userId") ?? ""
        userName = document.string("userName") ?? ""
        status = (document.string("status")).flatMap(BookingStatus.init(rawValue:)) ?? .waitlisted
        qrToken = document.string("qrToken") ?? ""
    }
}
